### PR TITLE
[pytorch][perf] Use NNPACK for strided convolutions.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Currently fp32 strided convolutions go through a code path that necessitates an im2col (THNN_Floatunfolded_copy) followed by GEMM (THFloatBlas_gemm).  Using NNPACK to achieve the same end goal (via compute_input_packing + compute_matrix_multiplication)
) seems to be more efficient and have better multi-threaded scaling.

All timings are on Pixel 2.

_Before_

Single-threaded: Main run finished. Milliseconds per iter: **353.043**. Iters per second: 2.83252

Multi-threaded: Main run finished. Milliseconds per iter: **361.032**. Iters per second: 2.76984

_After_

Single-threaded: Main run finished. Milliseconds per iter: **333.176**. Iters per second: 3.00141

Multi-threaded: Main run finished. Milliseconds per iter: **247.567**. Iters per second: 4.03931

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2029,7 +2029,7 @@
 - func: _nnpack_available() -> bool
   use_c10_dispatcher: full
 
-- func: _nnpack_spatial_convolution(Tensor input, Tensor weight, Tensor? bias, int[2] padding) -> Tensor
+- func: _nnpack_spatial_convolution(Tensor input, Tensor weight, Tensor? bias, int[2] stride, int[2] padding) -> Tensor
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/native_functions.yaml]: Who should I be adding to review this change?

File: test/test_nn.py [hunk 2]
```diff
@@ -8564,8 +8564,8 @@ def test_nnpack_conv(self):
                     weight = torch.randn(weight_shape, requires_grad=True, dtype=torch.float)
                     if has_bias:
                         bias = torch.randn([chan_out], requires_grad=True, dtype=torch.float)
-                    output = torch._nnpack_spatial_convolution(input, weight, padding=padding, bias=bias)
-                    output_expected = torch.nn.functional.conv2d(input, weight, padding=padding, bias=bias)
+                    output = torch._nnpack_spatial_convolution(input, weight, stirde=stride, padding=padding, bias=bias)
```
Comment (ajtulloch) [inline test/test_nn.py]: do these tests pass? `stirde` should be `stride` right?

File: aten/src/ATen/native/Convolution.cpp [hunk 3]
```diff
@@ -648,7 +647,7 @@ at::Tensor _convolution(
   } else if (input.device().type() == c10::DeviceType::CPU || input.device().type() == c10::DeviceType::CUDA) {
     if (params.use_cpu_depthwise3x3_winograd(input, weight)) {
       output = convolution_depthwise3x3_winograd_stub(
```
Comment (ajtulloch) [inline aten/src/ATen/native/Convolution.cpp]: Did this not work before?

**Review by ajtulloch — COMMENTED** [review]
Looks reasonable, minor q on the unrelated changes (and the typo?)

File: test/test_nn.py (hunk 2 above)
Comment (AshkanAliabadi) [reply test/test_nn.py]: Great find.  This should be at least part of the reason with these test failures.

File: aten/src/ATen/native/Convolution.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp]: This is absolutely nuts.  Here's how I understand it.

This is partly a case of double inversion where I was incorrectly passing the arguments as (padding, stride) at call site here while the function was expecting (stride, padding), and again incorrectly passing them swapped a second time as (stride, padding) to calculate_conv_output_size which expected its arguments as (padding, stride) so these canceled out!  So that's that.

Now stride was also not used in _convolution_depthwise3x3_winograd beyond this point since strided convolutions are not supported by the code path so didn't matter what incorrect value it had after this point, but padding was actually used after this point, and it contained whatever value stride had, namely 1, since we gate this call on the condition that stride is 1, but padding was also set to 1 for 3x3 kernels hence why things were working!!

I shudder at the thought of how many bugs of this nature are hidden in any software of non-trivial size. :)

**Review by AshkanAliabadi — COMMENTED** [review]
Thank you for your comments Andrew.  I tested this change on 5 phones and it seems to be a bit of hit and miss but mostly a hit hence why I kind of let it hanging here for a while.  There is definitely something going on with our multi-threaded codepath which make us perform poorly compared to Caffe 2 but the single-threaded gap is closing.  I've got a feeling the multi-threaded issue should be something silly and trivial, and am looking into that now.

**Comment by AshkanAliabadi** [comment]
Abandoning in favor of https://github.com/pytorch/pytorch/pull/27340 (exact same PR) and https://github.com/pytorch/pytorch/pull/27341 (multi-threaded improvements) as part of a ghstack.
